[5.5] stdlib: change the console to UTF-8 on start

compnerd opened the pull request:
This adjusts the Windows console to switch the codepage to UTF-8.  This
is important as the default codepage (CP437) does not allow for UTF-8
output, but expects ASCII.  However, strings in Swift are assumed to be
UTF-8, which means that there is now a conversion mismatch.

Because the console mode persists beyond the duration of the application
as it is state local to the console and not the C runtime, we should
restore the state of the console before termination.  We do this by
registering a termination handler via `atexit`.  This means that an
abnormal termination (e.g. via `fatalError`) will irrevocably alter the
state of the console (interestingly enough, `chcp` will still report the
original console codepage even though the console will internally be set
to UTF-8).

Fixes: SR-13807
(cherry picked from commit 9847cab8d3e5b74101a54bd3871f5002ffaf688c)

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves SR-NNNN.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd committed:
stdlib: change the console to UTF-8 on start

This adjusts the Windows console to switch the codepage to UTF-8.  This
is important as the default codepage (CP437) does not allow for UTF-8
output, but expects ASCII.  However, strings in Swift are assumed to be
UTF-8, which means that there is now a conversion mismatch.

Because the console mode persists beyond the duration of the application
as it is state local to the console and not the C runtime, we should
restore the state of the console before termination.  We do this by
registering a termination handler via `atexit`.  This means that an
abnormal termination (e.g. via `fatalError`) will irrevocably alter the
state of the console (interestingly enough, `chcp` will still report the
original console codepage even though the console will internally be set
to UTF-8).

Fixes: SR-13807
(cherry picked from commit 9847cab8d3e5b74101a54bd3871f5002ffaf688c)

compnerd commented:
@swift-ci please test

compnerd commented:
CC: @shahmishal @mikeash

shahmishal commented:
@compnerd 
```
Unexpectedly Passed Tests (1):
  Swift(windows-x86_64) :: Concurrency/Runtime/async_taskgroup_throw_rethrow.swift
```

We need to re-enable this test?

compnerd commented:
Hmm, that is intriguing ... it _is_ enabled on main.  I just am not sure what it picked up to now pass on 5.5.  It _might_ be the LLVM fix?

compnerd committed:
Update async_taskgroup_throw_rethrow.swift

Enable the test that was previously failing on 5.5.  This has been stable on main for a while, and seems to also be now passing on 5.5

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
Ah, yeah, seems that the other tests are similarly now passing.

compnerd committed:
Update async_taskgroup_asynciterator_semantics.swift

compnerd committed:
Update async_taskgroup_is_asyncsequence.swift

compnerd commented:
@swift-ci please test Windows platform

compnerd commented:
@swift-ci please test
